File: hooks/scripts/hamr_bypass_detector.py
"""HAMR bypass-detector for pretool_guard PreToolUse hook.

Refs Epic #2029 #2235 wiring follow-on. Python mirror of
scripts/global/hamr-bypass-detector.js (#2220) to avoid subprocess overhead
on every Bash invocation. Detector logic kept minimal; canonical regex set
matches the JS module per regression test.
"""

from __future__ import annotations

import json
import os
import re
import time
from pathlib import Path

FLEET_PORT = 11434
LOOPBACK_IP_PARTS = (127, 0, 0, 1)
DEFAULT_FLEET_IP_PARTS = (100, 91, 113, 16)

PAID_PROVIDER_REGEXES = [
    ("anthropic", re.compile(r"https?://api\.anthropic\.com", re.I)),
    ("openai", re.compile(r"https?://api\.openai\.com", re.I)),
    ("openrouter", re.compile(r"https?://openrouter\.ai", re.I)),
    ("gemini", re.compile(r"https?://generativelanguage\.googleapis\.com", re.I)),
    ("groq", re.compile(r"https?://api\.groq\.com", re.I)),
    ("cerebras", re.compile(r"https?://api\.cerebras\.ai", re.I)),
]

OVERRIDE_MARKER_RE = re.compile(r"#\s*hamr-bypass-ok:\s*(.+?)(?:\n|$)", re.I)
CURL_INVOKE_RE = re.compile(r"\b(?:curl|wget|http|httpie)\b", re.I)


def _ip_from_parts(parts):
    return ".".join(str(p) for p in parts)


def _fleet_substrings():
    fleet_ip = os.environ.get("FLEET_HOST_IP") or _ip_from_parts(DEFAULT_FLEET_IP_PARTS)
    loopback = _ip_from_parts(LOOPBACK_IP_PARTS)
    return [
        ("ollama-fleet", f"{fleet_ip}:{FLEET_PORT}"),
        ("ollama-local-ip", f"{loopback}:{FLEET_PORT}"),
        ("ollama-local-name", f"localhost:{FLEET_PORT}"),
    ]


def detect_bypass(cmd_string):
    text = str(cmd_string or "")
    if not CURL_INVOKE_RE.search(text):
        return {"detected": False, "reason": "not-http-invocation"}
    matches = []
    for name, regex in PAID_PROVIDER_REGEXES:
        if regex.search(text):
            matches.append({"name": name, "paid": True})
    for name, literal in _fleet_substrings():
        if literal in text:
            matches.append({"name": name, "paid": False})
    if not matches:
        return {"detected": False, "reason": "no-known-provider-url"}
    override_match = OVERRIDE_MARKER_RE.search(text)
    if override_match:
        return {"detected": True, "suppressed": True, "reason": "override-marker-present",
                "override_reason": override_match.group(1).strip(), "providers": matches}
    severity = "paid-bypass" if any(m["paid"] for m in matches) else "fleet-bypass"
    return {"detected": True, "suppressed": False, "providers": matches, "severity": severity}


def emit_incident(detection, incidents_path=None, now_ms=None):
    if not detection or not detection.get("detected") or detection.get("suppressed"):
        return None
    incidents_path = incidents_path or str(Path.home() / ".megingjord" / "incidents.jsonl")
    evt = {
        "ts": now_ms or int(time.time() * 1000),
        "version": 3,
        "service": "megingjord-hamr-bypass-detector",
        "env": "prod",
        "event": "hamr-bypass-detected",
        "pattern_id": "hamr-bypass-detected",
        "severity": detection["severity"],
        "providers": [p["name"] for p in detection["providers"]],
    }
    Path(incidents_path).parent.mkdir(parents=True, exist_ok=True)
    with open(incidents_path, "a", encoding="utf-8") as fh:
        fh.write(json.dumps(evt) + "\n")
    return evt

File: scripts/global/hamr-bypass-detector.js
'use strict';
// tier: 3
// hamr-bypass-detector — detect direct curls to known provider endpoints that bypass HAMR.
// Refs Epic #2029 #2220. Pure function; consumed by pretool_guard.py via subprocess or directly.

const FLEET_PORT = 11434;
const LOOPBACK_IP_PARTS = [127, 0, 0, 1];
const DEFAULT_FLEET_IP_PARTS = [100, 91, 113, 16];

const PAID_PROVIDER_REGEXES = [
  { name: 'anthropic', re: /https?:\/\/api\.anthropic\.com/i },
  { name: 'openai', re: /https?:\/\/api\.openai\.com/i },
  { name: 'openrouter', re: /https?:\/\/openrouter\.ai/i },
  { name: 'gemini', re: /https?:\/\/generativelanguage\.googleapis\.com/i },
  { name: 'groq', re: /https?:\/\/api\.groq\.com/i },
  { name: 'cerebras', re: /https?:\/\/api\.cerebras\.ai/i },
];

const OVERRIDE_MARKER_RE = /#\s*hamr-bypass-ok:\s*(.+?)(?:\n|$)/i;
const CURL_INVOKE_RE = /\b(?:curl|wget|http|httpie)\b/i;

function ipFromParts(parts) { return parts.join('.'); }

function getFleetSubstrings() {
  const fleetIp = process.env.FLEET_HOST_IP || ipFromParts(DEFAULT_FLEET_IP_PARTS);
  const loopback = ipFromParts(LOOPBACK_IP_PARTS);
  return [
    { name: 'ollama-fleet', literal: fleetIp + ':' + FLEET_PORT },
    { name: 'ollama-local-ip', literal: loopback + ':' + FLEET_PORT },
    { name: 'ollama-local-name', literal: 'localhost:' + FLEET_PORT },
  ];
}

function detectBypass(cmdString) {
  const text = String(cmdString || '');
  if (!CURL_INVOKE_RE.test(text)) return { detected: false, reason: 'not-http-invocation' };
  const matches = [];
  for (const provider of PAID_PROVIDER_REGEXES) {
    if (provider.re.test(text)) matches.push({ name: provider.name, paid: true });
  }
  for (const fleet of getFleetSubstrings()) {
    if (text.includes(fleet.literal)) matches.push({ name: fleet.name, paid: false });
  }
  if (matches.length === 0) return { detected: false, reason: 'no-known-provider-url' };
  const overrideMatch = text.match(OVERRIDE_MARKER_RE);
  if (overrideMatch) {
    return { detected: true, suppressed: true, reason: 'override-marker-present',
      override_reason: overrideMatch[1].trim(), providers: matches };
  }
  return { detected: true, suppressed: false, providers: matches,
    severity: matches.some(function (p) { return p.paid; }) ? 'paid-bypass' : 'fleet-bypass' };
}

function emitIncident(detection, incidentsPath, now) {
  if (!detection || !detection.detected || detection.suppressed) return null;
  const fs = require('node:fs');
  const path = require('node:path');
  const evt = {
    ts: now || Date.now(), version: 3,
    service: 'megingjord-hamr-bypass-detector', env: 'prod',
    event: 'hamr-bypass-detected', pattern_id: 'hamr-bypass-detected',
    severity: detection.severity,
    providers: detection.providers.map(function (provider) { return provider.name; }),
  };
  fs.mkdirSync(path.dirname(incidentsPath), { recursive: true });
  fs.appendFileSync(incidentsPath, JSON.stringify(evt) + '\n');
  return evt;
}

module.exports = { detectBypass, emitIncident, PAID_PROVIDER_REGEXES,
  OVERRIDE_MARKER_RE, getFleetSubstrings, FLEET_PORT };
